### PR TITLE
support distribute for fs

### PR DESCRIPTION
**Pull request by irisdingbj**

### CODE REVIEW

**Comment by ozevren** [comment]
Can you add tests?

**Comment by irisdingbj** [comment]
@ozevren Testing added and fixed some test failure. Please review

**Comment by ayj** [comment]
@irisdingbj , to fix the failing required e2e tests you'll need to rebase to pick up https://github.com/istio/istio/pull/9637.

**Comment by irisdingbj** [comment]
@ayj rebased, please review. thanks

File: galley/pkg/fs/fssource_test.go [hunk 1]
```diff
@@ -326,23 +328,63 @@ func TestFsSource_DeletePartResorceInFile(t *testing.T) {
...
+	}
+	d := runtime.NewInMemoryDistributor()
+	processor := runtime.NewProcessor(s, d)
+	err = processor.Start()
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
 
+	// Wait "long enough"- the default publishing timing is 1 second
+	time.Sleep(time.Second * 2)
```
[the remark this answers is not in the document]
Comment (ozevren) [reply galley/pkg/fs/fssource_test.go]: Let me rephrase that: Do not use sleeps without any feedback loop. It is ok to use a sleep/retry pattern to coordinate, but a single sleep followed by a pass/fail lookup is recipe for flakiness.

You can do a sleep/retry, or (even better) you can add some sort of channel/publishing support to the in-memory distributor, so that you can wait on distribution. e.g.:
```go
func (s* InMemorySnapshot) ListenChanges(chan cancel, ListenerFn fn) {
   s.listenerMutex.Lock()
   snapshotChan := make(chan Snapshot, 1)
   s.listeners = append(s.listeners, listenerEntry{fn, snapshotChan})
   s.listenerMutex.Unlock()
   for {
     select {
       case <-cancel:
         close(snapshotChan)
         return
       case s <- snapshotChan:
          fn(s)
    }
  }
}
// Called internally when a snapshot is set
func (s *InMemorySnapshot)  notifyListeners() {
   s.listenerMutex.Lock()
   defer s.listenerMutex.Unlock()
   for _, e := range s.listeners {
      e.snapshotChan <= s.snapshot
   }
}
```

[the remark this answers is not in the document]
Comment (irisdingbj) [reply galley/pkg/fs/fssource_test.go]: Thanks Oz, fixed now.

**Comment by irisdingbj** [comment]
@ozevren ^^^

File: galley/pkg/runtime/distributor.go [hunk 2]
```diff
@@ -57,3 +70,40 @@ func (d *InMemoryDistributor) ClearSnapshot(name string) {
...
+		return s
+	}
+	return nil
+}
+
+// ListenChanges registered listener and start listening snapshot changes in the distributor
+func (d *InMemoryDistributor) ListenChanges(cancel chan bool, fn ListenerFn) {
+	d.listenersLock.Lock()
+	snapshotChan := make(chan sn.Snapshot, 1)
+	d.listeners = append(d.listeners, &listenerEntry{fn, snapshotChan})
```
[the remark this answers is not in the document]
Comment (irisdingbj) [reply galley/pkg/runtime/distributor.go]: thanks Oz, fixed now
